go/pkg/bertyprotocol/api_attachment.go

```diff
@@ -0,0 +1,327 @@
+package bertyprotocol
+
+import (
+	"fmt"
+	"io"
+	"math/big"
+
+	ipfscid "github.com/ipfs/go-cid"
+	ipfsfiles "github.com/ipfs/go-ipfs-files"
+	ipfspath "github.com/ipfs/interface-go-ipfs-core/path"
+	libp2pcrypto "github.com/libp2p/go-libp2p-core/crypto"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/chacha20poly1305"
+
+	"berty.tech/berty/v2/go/pkg/bertytypes"
+	"berty.tech/berty/v2/go/pkg/errcode"
+)
+
+const (
+	attachmentCipherblockSize = 64 * 1024
+	attachmentPlainblockSize  = attachmentCipherblockSize - 16 // TODO: check if the overhead is really fixed
+	attachmentNonceIV         = 0
+	attachmentKeyV0Prefix     = "/libp2psk+xchacha20poly1305_64_0/" // TODO: replace when multikey rolls out
+)
+
+func (s *service) PrepareAttachment(stream ProtocolService_PrepareAttachmentServer) error {
+	headerMsg, err := stream.Recv()
+	if err != nil {
+		return fmt.Errorf("receive header: %s", err.Error())
+	}
+	if len(headerMsg.GetBlock()) > 0 {
+		return fmt.Errorf("unexpected non-empty block")
+	}
+	if headerMsg.GetHeader().GetDisableEncryption() {
+		return fmt.Errorf("disable_encryption not implemented")
+	}
+	s.logger.Debug("header received")
+
+	plaintext := attachmentPlaintextReader(stream, s.logger)
+	defer plaintext.Close()
+
+	sk, ciphertext, err := attachmentSealer(plaintext, s.logger)
+	s.logger.Debug("stream cipher init done", zap.Error(err))
+	if err != nil {
+		return fmt.Errorf("failed to open attachment stream: %s", err.Error())
+	}
+	defer ciphertext.Close()
+
+	ipfsPath, err := s.ipfsCoreAPI.Unixfs().Add(stream.Context(), ipfsfiles.NewReaderFile(ciphertext))
```

Comment:
I'm not sure, is this enough to pin the content? (I didn't check IPFS doc)

Reply:
pin is the default but @Jorropo  says that it's not guaranteed in the future, will force it